inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_bs_fs_yx_bsv16_fsv16.cl

```diff
@@ -0,0 +1,183 @@
+                    IN_VEC16 ch16_data = AS_TYPE(IN_VEC16, int_data);
+                    __attribute__((opencl_unroll_hint(16)))
+                    for (uint z = 0; z < 16; z++)
+                        result[z] = FUNC_CALL(apply_pooling)(result[z], (int)ch16_data[z]);
+#ifdef DYNAMIC_KERNEL_DIVIDER
+                    num_elements++;
+#endif
+                }
+            }
+        }
+    }
+
+#ifdef DYNAMIC_WITH_PADDING_KERNEL_DIVIDER
+    const int hend = min(offset_y + POOL_SIZE_Y, INPUT0_SIZE_Y + PADDING_SIZE_Y);
+    const int wend = min(offset_x + POOL_SIZE_X, INPUT0_SIZE_X + PADDING_SIZE_X);
+    const uint num_elements = (hend - offset_y) * (wend - offset_x);
+#endif
+#else  // !CHECK_BOUNDRY
+    uint input_idx = GET_DATA_BS_FS_YX_BSV16_FSV16_INDEX(INPUT0, b, f, offset_y, offset_x);
+    __attribute__((opencl_unroll_hint(POOL_SIZE_Y)))
+    for (uint j = 0; j < POOL_SIZE_Y; j++) {
+        __attribute__((opencl_unroll_hint(POOL_SIZE_X)))
+        for (uint i = 0; i < POOL_SIZE_X; i++) {
+                int4 int_data = vload4(0, (__global int*)(input + input_idx));
+                IN_VEC16 ch16_data = AS_TYPE(IN_VEC16, int_data);
+                __attribute__((opencl_unroll_hint(16)))
+                for (uint z = 0; z < 16; z++)
+                    result[z] = FUNC_CALL(apply_pooling)(result[z], (int)ch16_data[z]);
+               
+            input_idx += input_x_pitch;
+        }
+        input_idx += (input_y_pitch - POOL_SIZE_X * input_x_pitch);
+    }
+
+#if defined(DYNAMIC_KERNEL_DIVIDER) || defined(DYNAMIC_WITH_PADDING_KERNEL_DIVIDER)
+    const uint num_elements = POOL_SIZE_X * POOL_SIZE_Y;
+#endif
+#endif
+#if defined AVG_POOLING
+#if ENABLE_ROUND
+    int16 pool_result;
+    __attribute__((opencl_unroll_hint(16)))
+    for (uint i = 0; i < 16; i++) {
+#if defined(DYNAMIC_KERNEL_DIVIDER) || defined(DYNAMIC_WITH_PADDING_KERNEL_DIVIDER)
+        result[i] = convert_int(round(((float)result[i] / max(num_elements, (uint)1))));
+#else
+        result[i] = convert_int(round((float)result[i] / (int)(POOL_SIZE_Y * POOL_SIZE_X)));
+#endif
+    }
+#else
+    float16 pool_result;
+    __attribute__((opencl_unroll_hint(16)))
+    for (uint i = 0; i < 16; i++) {
+#if defined(DYNAMIC_KERNEL_DIVIDER) || defined(DYNAMIC_WITH_PADDING_KERNEL_DIVIDER)
+        pool_result[i] = (float)result[i] / max(num_elements, (uint)1);
+#else
+        pool_result[i] = (float)result[i] / (int)(POOL_SIZE_Y * POOL_SIZE_X);
+#endif
+    }
+#endif  // ENABLE_ROUND
+#else   // AVG_POOLING
+    int16 pool_result;
+    __attribute__((opencl_unroll_hint(16)))
+    for (uint i = 0; i < 16; ++i) {
+        pool_result[i] = result[i];
+    }
+#endif  // AVG_POOLING
+    OUT_VEC16 final_result = (OUTPUT_TYPE)(0);
+#if HAS_FUSED_OPS
+    //FUSED_OPS_LOAD_PER_SCALE
+    FUSED_OPS_PRELOAD
+#endif
+    __attribute__((opencl_unroll_hint(16)))
+    for (uint i = 0; i < 16; ++i) {
+#if HAS_FUSED_OPS
+        
+        FUSED_OPS;
```

Comment:
I suppose you should use here such construction as you use in other kernels of this commit.

#if HAS_FUSED_OPS && FUSED_OPS_CAN_USE_PRELOAD
        FUSED_OPS_PRELOAD
#endif

...
#if HAS_FUSED_OPS
            #if FUSED_OPS_CAN_USE_PRELOAD
                FUSED_OPS_CALC
            #else
                FUSED_OPS
            #endif
#endif

Reply:
Makes sense to me as well. @leszczyn what do you think?

Reply:
if it will  provide correct results, why not

Reply:
Changed as proposed.